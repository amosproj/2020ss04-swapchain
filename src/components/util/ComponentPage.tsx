import React from "react"

interface Props {
  title?: string
  children: any
  bg?: string
}

export const ComponentPage = (props: Props) => {
  return (
    <div className="container mx-auto">
      <main className={`pb-10 mx-auto ${props.bg} ${props.bg ? "shadow-sm" : ""}  p-4 sm:px-6 lg:px-8`}>
        <h1 className="py-4 text-3xl font-semibold leading-tight text-gray-900">{props.title}</h1>
        {props.children}
      </main>
    </div>
  )
}
