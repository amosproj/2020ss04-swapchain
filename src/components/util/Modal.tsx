import React from "react"
import { ReactComponent as X } from "../../icons/x.svg"
type Props = {
  close: () => void
  open: boolean
  children: any
  title?: string
}

export const Modal = (props: Props) => {
  return (
    <div className={props.open ? "" : "hidden"}>
      <div className="absolute inset-0 bg-gray-500 opacity-75"></div>
      <div className="absolute inset-0 z-10 flex items-center justify-center">
        <div>
          <header className="flex items-center justify-between">
            <h1 className="text-lg font-semibold text-gray-500">{props.title}</h1>
            <X className="w-8 h-8 text-gray-400 duration-200 ease-in-out hover:text-gray-800" onClick={props.close}></X>
          </header>
          <div className="flex flex-col items-center justify-around flex-grow md:flex-row">{props.children}</div>
        </div>
      </div>
    </div>
  )
}
