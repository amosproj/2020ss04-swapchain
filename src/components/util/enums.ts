export enum State {
  IDLE,
  RUNNING,
  SUCCESS,
  ERROR,
  FAILURE,
}

export enum Currency {
  BTC = "BTC",
  BTS = "BTS",
}

export enum Network {
  MAINNET = "mainnet",
  TESTNET = "testnet",
}

export enum Priority {
  HIGH,
  MEDIUM,
  LOW,
}
