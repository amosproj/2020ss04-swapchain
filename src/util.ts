import * as bitcoin from "bitcoinjs-lib"

export const fakeKey = (length: number, network?: string) => {
  let prefix = ""
  if (network === "mainnet") {
    prefix = "L"
  } else if (network === "testnet") {
    prefix = "c"
  }

  return prefix + hash(Math.random().toString()).substring(0, length)
}

export const hash = (s: string): string => {
  return bitcoin.crypto.hash256(Buffer.from(s)).toString("hex")
}

export const toPublicKey = (privateKey: string): string => {
  const keyPair = bitcoin.ECPair.fromWIF(privateKey, bitcoin.networks.testnet)
  const keyPairCompressed = bitcoin.ECPair.fromPrivateKey(keyPair.privateKey!, { compressed: true })
  return keyPairCompressed.publicKey.toString("hex")
}
