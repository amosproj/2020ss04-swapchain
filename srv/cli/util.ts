import readline from "readline"

/**
 * Get input from stdin
 *
 * @param str - A question to ask the user.
 * @returns answer from the user
 */
export async function askUser(str: string): Promise<string> {
  const read = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  })

  return new Promise((resolve) => read.question("\u001b[36;1m" + str + "\u001b[0m", resolve))
}
