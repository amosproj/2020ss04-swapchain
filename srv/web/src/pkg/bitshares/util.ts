import { Apis as btsWebsocketApi, ChainConfig } from "bitsharesjs-ws"
import { PrivateKey } from "bitsharesjs"

export async function getAccount(privateKey: string, node: string, network: string): Promise<string> {

    if (privateKey.length !== 51) {
        throw new Error("Invalid private key length.")
    }

    // Set chain IDs to get correct prefix for public key
    if (network === "mainnet") {
        ChainConfig.setChainId("4018d7844c78f6a6c41c6a552b898022310fc5dec06da467ee7905a8dad512c8")
    } else if (network === "testnet") {
        ChainConfig.setChainId("39f5e2ede1f8bc1a3a54a7914414e3779e33193f1f5693510e73cb7a87617447")
    } else {
        throw new Error("Invalid network name. Choose mainnet or testnet.")
    }

    const publicKey = PrivateKey.fromWif(privateKey).toPublicKey().toString()

    await btsWebsocketApi.instance(node, true).init_promise

    const accounts = await btsWebsocketApi.db.get_key_references([publicKey])

    if (typeof accounts[0][0] === "undefined") {
        throw new Error(`Could not find any accounts for private key ${privateKey} on ${network}`)
    }

    const accountID = accounts[0][0]

    const accountNames = await btsWebsocketApi.db.get_accounts([accountID])

    if (Object.keys(accountNames[0]).length === 0) {
        throw new Error(`Could not get accounts for account ID ${accountID} on ${network}`)
    }
    
    return accountNames[0].name
}