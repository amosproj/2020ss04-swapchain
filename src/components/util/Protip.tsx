import React from "react"
import { ReactComponent as LightBulb } from "../../icons/light-bulb.svg"
import { Network } from "../util/enums"

interface Props {
  network: Network
}

export const Protip = (props: Props) => {
  return (
    <div className="flex flex-col items-center px-3 py-2 space-x-1 md:flex-row">
      <div className="flex items-center">
        <LightBulb className="h-6 text-gray-600"></LightBulb>
        <span className="font-semibold text-gray-900">ProTip!</span>
      </div>
      <div className="flex flex-col items-center md:flex-row">
        <span className="text-center text-gray-700">Check your transaction on the blockchain:</span>
        <div className="flex px-2 space-x-2">
          <a
            target="blank"
            className="text-blue-400"
            href={props.network === Network.MAINNET ? "https://blockstream.info/" : "https://blockstream.info/testnet/"}
          >
            Bitcoin
          </a>
          <a target="blank" className="text-blue-400" href="https://wallet.bitshares.org/#/">
            Bitshares
          </a>
        </div>
      </div>
    </div>
  )
}
