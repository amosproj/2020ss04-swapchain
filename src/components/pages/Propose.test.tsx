import React from "react"
import { Propose } from "./Propose"
import renderer from "react-test-renderer"
import * as util from "../../pkg/util/util"
jest.mock("../../pkg/util/util")

jest.spyOn(util, "fakeKey").mockReturnValue("randomLookingHash")
it("renders correctly", () => {
  const tree = renderer.create(<Propose></Propose>).toJSON()
  expect(tree).toMatchSnapshot()
})
