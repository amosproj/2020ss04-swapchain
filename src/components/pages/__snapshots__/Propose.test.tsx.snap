// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="flex flex-col items-center bg-white border rounded shadow-lg"
>
  <div
    className="flex flex-col justify-center p-4 space-y-4"
  >
    <div
      className="p-8"
    >
      <div>
        <h2
          className="text-3xl font-semibold leading-tight text-gray-800 "
        >
          Propose a new Atomic Cross Chain Swap
        </h2>
        <div
          className="flex flex-col mt-12 space-y-10"
        >
          <section>
            <label
              className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
            >
              On what network do you want to trade
            </label>
            <div
              className="flex flex-col justify-center space-y-4 md:flex-row md:space-x-4 md:space-y-0"
            >
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded bg-gray-100 hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-600"
                    >
                      Mainnet
                    </span>
                    <span
                      className="text-xs font-thin text-gray-400"
                    />
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-400"
                  >
                    You are sending real money!
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-300"
                  >
                    <svg
                      className="h-8 text-red-600 hidden"
                    >
                      exclamation.svg
                    </svg>
                  </span>
                </div>
              </button>
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded border-teal-400 shadow hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-900"
                    >
                      Testnet
                    </span>
                    <span
                      className="text-xs font-thin text-gray-600"
                    />
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-600"
                  >
                    Send test currencies on the testnet
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-900 "
                  />
                </div>
              </button>
            </div>
          </section>
          <section>
            <label
              className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
            >
              What do you want to trade
            </label>
            <div
              className="flex flex-col items-center space-y-4 md:flex-row md:space-x-4 md:space-y-0"
            >
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded border-teal-400 shadow hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-900"
                    >
                      Bitcoin
                    </span>
                    <span
                      className="text-xs font-thin text-gray-600"
                    />
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-600"
                  >
                    You are giving Bitcoin away
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-900 "
                  >
                    BTC
                  </span>
                </div>
              </button>
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded bg-gray-100 hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-600"
                    >
                      Bitshares
                    </span>
                    <span
                      className="text-xs font-thin text-gray-400"
                    />
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-400"
                  >
                    You are giving Bitshares away
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-300"
                  >
                    BTS
                  </span>
                </div>
              </button>
            </div>
          </section>
          <section>
            <h2
              className="text-xl font-semibold leading-tight text-gray-800"
            >
              Amount
            </h2>
            <div
              className="items-center justify-between mt-4 md:space-x-4 md:flex"
            >
              <div
                className="flex-grow"
              >
                <label
                  className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
                >
                  Amount to send
                </label>
                <div
                  className="relative"
                >
                  <span
                    className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600"
                  >
                    BTC
                  </span>
                  <input
                    className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                    min={0}
                    name="amountToSend"
                    onChange={[Function]}
                    placeholder="0.00000000"
                    required={true}
                    step={1e-8}
                    type="number"
                    value=""
                  />
                </div>
              </div>
              <div
                className="flex-grow"
              >
                <label
                  className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
                >
                  Rate
                </label>
                <div
                  className="relative"
                >
                  <div
                    className="absolute inset-y-0 right-0 flex flex-col items-center justify-center mr-6 text-xs text-gray-600"
                  >
                    <span>
                      BTS
                    </span>
                    <span
                      className="border-t border-gray-500"
                    >
                      BTC
                    </span>
                  </div>
                  <input
                    className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                    min={0}
                    name="rate"
                    onChange={[Function]}
                    placeholder="0.0000"
                    required={true}
                    step={1e-8}
                    type="number"
                    value=""
                  />
                </div>
              </div>
              <div
                className="flex-grow"
              >
                <label
                  className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
                >
                  excluding fees you will receive
                </label>
                <div
                  className="relative"
                >
                  <span
                    className="absolute inset-y-0 right-0 flex items-center mr-6 text-sm text-gray-600"
                  >
                    BTS
                               
                  </span>
                  <span
                    className="block w-full py-3 font-mono text-center text-gray-700 border border-gray-200 rounded focus:border-teal-500"
                  >
                    0
                  </span>
                </div>
              </div>
            </div>
          </section>
          <section>
            <h2
              className="text-xl font-semibold leading-tight text-gray-800"
            >
              Your Data
            </h2>
            <div
              className="flex flex-col space-y-4"
            >
              <div
                className="items-center justify-between mt-4 md:space-x-2 md:flex"
              >
                <div
                  className="md:w-1/2"
                >
                  <label
                    className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
                  >
                    Bitcoin private key
                  </label>
                  <input
                    className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                    name="bitcoinPrivateKey"
                    onChange={[Function]}
                    placeholder="randomLookingHash"
                    required={true}
                    type="text"
                    value=""
                  />
                </div>
                <div
                  className="md:w-1/2"
                >
                  <label
                    className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
                  >
                    Bitshares private key
                  </label>
                  <input
                    className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                    name="bitsharesPrivateKey"
                    onChange={[Function]}
                    placeholder="5randomLookingHash"
                    required={true}
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <p
                className="px-4 mx-auto text-sm text-center text-gray-500"
              >
                Your public Bitcoin key and Bitshares account name can be derived from your private key. Your private keys will never leave your browser, they are only used to sign your transactions. 
                <a
                  className="relative text-xs text-blue-500"
                  href="/"
                >
                  Read more in our docs.
                </a>
              </p>
            </div>
            <div
              className="mt-2 "
            >
              <label
                className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
              >
                Bitcoin transaction ID to spend
              </label>
              <input
                className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                name="bitcoinTxID"
                onChange={[Function]}
                placeholder="randomLookingHash"
                required={true}
                type="text"
                value=""
              />
            </div>
          </section>
          <section>
            <h2
              className="text-xl font-semibold leading-tight text-gray-800"
            >
              Counterparty Data
            </h2>
            <div
              className="items-center justify-between mt-4 md:space-x-4 md:flex"
            >
              <div
                className="md:w-1/2"
              >
                <label
                  className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
                >
                  Bitcoin public key
                </label>
                <input
                  className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                  name="counterpartyBitcoinPublicKey"
                  onChange={[Function]}
                  placeholder="02randomLookingHash"
                  required={true}
                  type="text"
                  value=""
                />
              </div>
              <div
                className="md:w-1/2"
              >
                <label
                  className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
                >
                  Bitshares account name
                </label>
                <input
                  className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
                  name="counterpartyBitsharesAccountName"
                  onChange={[Function]}
                  placeholder="amos"
                  required={true}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </section>
          <section>
            <label
              className="text-xs leading-4 tracking-wider text-gray-700 uppercase"
            >
              Choose your priority
            </label>
            <div
              className="flex flex-col space-y-4 md:flex-row md:space-x-4 md:space-y-0"
            >
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded border-teal-400 shadow hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-900"
                    >
                      HIGH
                    </span>
                    <span
                      className="text-xs font-thin text-gray-600"
                    />
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-600"
                  >
                    You pay the highest fees to increase the chance for your transaction to be picked up by the miners.
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-900 "
                  />
                </div>
              </button>
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded bg-gray-100 hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-600"
                    >
                      MEDIUM
                    </span>
                    <span
                      className="text-xs font-thin text-gray-400"
                    />
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-400"
                  >
                    You pay a moderate amount of fees so miners will probably confirm your transaction soon.
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-300"
                  />
                </div>
              </button>
              <button
                className="focus:outline-none flex w-full items-center justify-between border rounded bg-gray-100 hover:bg-white hover:shadow-sm"
                onClick={[Function]}
              >
                <div
                  className="flex flex-col items-start justify-between p-4"
                >
                  <div
                    className="flex items-center justify-between w-full"
                  >
                    <span
                      className="text-left font-semibold text-gray-600"
                    >
                      LOW
                    </span>
                    <span
                      className="text-xs font-thin text-gray-400"
                    />
                  </div>
                  <p
                    className="mt-2 text-sm text-left text-gray-400"
                  >
                    You pay the lowest fees but might have to wait a few more blocks for your transaction to be confirmed.
                  </p>
                </div>
                <div
                  className="flex items-center justify-center"
                >
                  <span
                    className="mr-4 text-3xl font-semibold hidden md:block text-gray-300"
                  />
                </div>
              </button>
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
  <footer
    className="flex justify-center w-full p-4 bg-gray-100 border-t border-gray-200"
  >
    <div>
      <button
        className="px-8 py-3 font-semibold text-gray-900 uppercase transition duration-200 ease-in-out bg-white border rounded-sm shadow-sm focus:outline-none hover:border-teal-500 hover:bg-white border-teal-400"
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </div>
  </footer>
</div>
`;
